bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,80 +1,302 @@
+        }
+
+        let failed_attempts = self.num_retries();
+
+        let worker_status = match self.status_client.status(StatusRequest {}).await {
+            Ok(response) => response.into_inner(),
+            Err(e) => {
+                error!("Failed to check worker status ({}): {}", self.address(), e);
+                self.set_health_status(WorkerHealthStatus::Unhealthy {
+                    num_failed_attempts: failed_attempts + 1,
+                    first_fail_timestamp: Instant::now(),
+                    reason: e.message().to_string(),
+                });
+                return;
+            },
+        };
+
+        if worker_status.version.is_empty() {
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker version is empty".to_string(),
+            });
+            return;
+        }
+
+        if !self.is_valid_version(&worker_status.version) {
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker version is invalid".to_string(),
+            });
+            return;
+        }
+
+        self.version = worker_status.version;
+
+        let worker_supported_proof_type =
+            match ProverType::try_from(worker_status.supported_proof_type) {
+                Ok(proof_type) => proof_type,
+                Err(e) => {
+                    error!(
+                        "Failed to convert worker supported proof type ({}): {}",
+                        self.address(),
+                        e
+                    );
+                    self.set_health_status(WorkerHealthStatus::Unhealthy {
+                        num_failed_attempts: failed_attempts + 1,
+                        first_fail_timestamp: Instant::now(),
+                        reason: e.to_string(),
+                    });
+                    return;
+                },
+            };
+
+        if supported_prover_type != worker_supported_proof_type {
+            self.set_health_status(WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: format!("Unsupported prover type: {}", worker_supported_proof_type),
+            });
+            return;
+        }
+
+        self.set_health_status(WorkerHealthStatus::Healthy);
+    }
+
+    /// Sets the worker availability.
+    pub fn set_availability(&mut self, is_available: bool) {
+        self.is_available = is_available
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of retries the worker has had.
+    pub fn num_retries(&self) -> usize {
```

Comment:
nit: I think I'd still name this something that reflected the fact that these are not retries, but rather how many failed attempts there were overall (expanded [here](https://github.com/0xPolygonMiden/miden-base/pull/1287#discussion_r2045228856))